### ISSUE TITLE
in method 'swt', argument 1 of type 'char *'

### ISSUE THREAD

**Issue by nerdykamil** [marrrcin/swt-python#1]
When i try to use the wrapper on my image it gives the error;
_**in_ method 'swt', argument 1 of type 'char *'**_
i have tried passing it as string, it works but there is no output.
Any leads?

**Comment by marrrcin** [comment]
Do you pass your image to SWT wrapper like this?
```python
image_name = "test_input.jpg"
bytes = open(image_name, "rb").read()
swt_result_raw = ccvwrapper.swt(bytes, len(bytes), 1024, 1360) # width = 1024, height = 1360
swt_result = np.reshape(swt_result_raw, (len(swt_result_raw) / 4, 4))
```

**Comment by nerdykamil** [comment]
yes, exactly like that. Just that I have used python3.5 when compiling c code

**Comment by maky-hnou** [comment]
Same problem with python 3.6!

**Comment by UdovenkoVolodymyr** [comment]
Same problem with python 3.6!

**Comment by justiniggy** [comment]
I meet the same problem. Any idea about implementing these code in python3.7? Thanks a lot

**Comment by OppoOskar** [comment]
So, spent 5hours trying to get this to work in Python 3.6 and finally made it!

Just to inform you, I DO NOT know SWIG and only have a basic understanding of C. However after trial and error I came up with this method, it is far from perfect but it works.

Firstly in **ccvwrapper.i** the following addition is needed:
```
%typemap(in) (char *bytes, int array_length) {
    Py_ssize_t len;
    PyBytes_AsStringAndSize($input, &$1, &len);
    $2 = (int)len;
}
```

Then in **main.py** we need to make some changes aswell.
```
from __future__ import print_function
import ccvwrapper
import numpy as np
from skimage import draw
from skimage.io import imread, imshow, imsave
#from matplotlib import pyplot as plt
import matplotlib
matplotlib.use('agg')
import matplotlib.pyplot as plt


def rectangle_perimeter(r0, c0, width, height, shape=None, clip=False):
    rr, cc = [r0, r0 + width, r0 + width, r0], [c0, c0, c0 + height, c0 + height]

    return draw.polygon_perimeter(rr, cc, shape=shape, clip=clip)


if __name__ == "__main__":
    image_name = "instrumentB.jpg"
    bytes = open(image_name, "rb").read()
    swt_result_raw = ccvwrapper.swt(bytes, 1024, 1360) #NEW
    swt_result_array = np.asarray(swt_result_raw).astype(int)

    swt_result = np.reshape(swt_result_array, (len(swt_result_array) / 4, 4))

    image = imread(image_name, as_gray=False)
    for x, y, width, height in swt_result:
        for i in range(0, 3): # just to make lines thicker
            rr, cc = rectangle_perimeter(y + i, x + i, height, width, shape=image.shape, clip=True)
            image[rr, cc] = (255, 0, 0)

    #imshow(image)
    imsave("result.jpg", image)
    plt.show()
```
The main part of the code I changed is that the swt command now only receives 3 commands and we have to convert the list received from C to a numpy array.

Matplotlib kept complaining about "tkinter" and could not figure out how to solve it, so switched to "aggr", a non-GUI version of matlibplot. So I can no longer use imshow, instead I use imsave.

Numpy kept complaining about floats, so I had to downgrade to numpy version 1.11.0.

As I said, I got it working for me but I really do not have any explanation as to why it works. I just hope this can help you guys get it working aswell, good luck!

**Comment by marrrcin** [comment]
Thanks for dealing with this. I've planned to revisit my blog post soon and update it for modern python.

**Comment by leehaining** [comment]
I changed the procedure according to your suggestion,But there is another problem：
File "/home/hn/swt-python-master/main.py", line 22, in <module>
    swt_result_raw = ccvwrapper.swt(bytes, 1024, 1360)
TypeError: swt() takes exactly 4 arguments (3 given)

**Comment by OppoOskar** [comment]
@leehaining I found my old files, apparently, I had a different build.sh as well. I haven't looked into it so it might just have been a small cleanup (might not solve your problem). 

I created a fork of the repository and uploaded it here:
[https://github.com/OppoOskar/swt-python](url)

If @marrrcin wants it removed I will take it down but for now, I put it there.

This is not a solution I actively use or support and I haven't tested it since I used it in 2019 so use it at your own risk.
